### PR TITLE
chore: convert extension apis to TypeScript

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Converts extensions-related files to TS.

cc @MarshallOfSound 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: lib/renderer/extensions/event.ts [hunk 1]
```diff
@@ -1,26 +1,22 @@
...
-  removeListener (callback) {
+  removeListener (callback: Function) {
     const index = this.listeners.indexOf(callback)
     if (index !== -1) {
       this.listeners.splice(index, 1)
     }
   }
 
-  emit (...args) {
+  emit (...args: any) {
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/event.ts]:
```suggestion
  emit (...args: any[]) {
```

File: lib/renderer/extensions/event.ts [hunk 2]
```diff
@@ -1,26 +1,22 @@
...
     for (const listener of this.listeners) {
       listener(...args)
     }
   }
 }
 
-module.exports = Event
+export default Event
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/event.ts]: Can we prefer named exports for things not in `*/api/*`.  E.g. this should be `export class Event`

File: lib/renderer/extensions/i18n.ts [hunk 3]
```diff
@@ -1,59 +1,58 @@
...
+  content: string;
+  example?: string;
+}
 
-const getMessages = (extensionId) => {
+const getMessages = (extensionId: number) => {
   try {
     const data = ipcRendererUtils.invokeSync('CHROME_GET_MESSAGES', extensionId)
-    return JSON.parse(data) || {}
+    return JSON.parse(data as any) || {}
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/i18n.ts]:
```suggestion
    return JSON.parse(data) || {}
```

File: lib/renderer/extensions/i18n.ts [hunk 4]
```diff
@@ -1,59 +1,58 @@
...
+
+interface Placeholder {
+  content: string;
+  example?: string;
+}
 
-const getMessages = (extensionId) => {
+const getMessages = (extensionId: number) => {
   try {
     const data = ipcRendererUtils.invokeSync('CHROME_GET_MESSAGES', extensionId)
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/i18n.ts]:
```suggestion
    const data = ipcRendererUtils.invokeSync<string>('CHROME_GET_MESSAGES', extensionId)
```

File: lib/renderer/extensions/i18n.ts [hunk 5]
```diff
@@ -1,59 +1,58 @@
...
+const replacePlaceholders = (message: string, placeholders: Record<string, Placeholder>, substitutions: string[] | string) => {
+  if (typeof substitutions === 'string') substitutions = [substitutions]
+  if (!Array.isArray(substitutions)) substitutions = []
 
   if (placeholders) {
-    Object.keys(placeholders).forEach((name) => {
+    Object.keys(placeholders).forEach((name: string) => {
       let { content } = placeholders[name]
-      content = replaceNumberedSubstitutions(content, substitutions)
+      content = replaceNumberedSubstitutions(content, substitutions as string[])
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/i18n.ts]: This `as string[]` shouldn't be required, it should be known at this point that `substitutions` is definitely an array.🤔 In order to prove it you may need to make a `const` redef

```js
const substitutionsArray = Array.isArray(substitutions) ? substitutions : []
```

Which is const and of type `string[]`

File: lib/renderer/extensions/storage.ts [hunk 6]
```diff
@@ -1,12 +1,10 @@
-'use strict'
+import * as ipcRendererUtils from '@electron/internal/renderer/ipc-renderer-internal-utils'
 
-const ipcRendererUtils = require('@electron/internal/renderer/ipc-renderer-internal-utils')
-
-const getStorage = (storageType, extensionId, callback) => {
+const getStorage = (storageType: string, extensionId: number, callback: Function) => {
   if (typeof callback !== 'function') throw new TypeError('No callback provided')
 
   ipcRendererUtils.invoke('CHROME_STORAGE_READ', storageType, extensionId)
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/storage.ts]:
```suggestion
  ipcRendererUtils.invoke<string>('CHROME_STORAGE_READ', storageType, extensionId)
```

File: lib/renderer/extensions/storage.ts [hunk 7]
```diff
@@ -1,12 +1,10 @@
...
-    .then(data => {
+    .then((data: any) => {
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/storage.ts]:
```suggestion
    .then((data) => {
```

File: lib/renderer/extensions/storage.ts [hunk 8]
```diff
@@ -58,37 +56,32 @@ const getStorageManager = (storageType, extensionId) => {
...
     },
 
-    clear (callback) {
+    clear (callback: Function) {
       setStorage(storageType, extensionId, {}, callback)
     }
   }
 }
 
 module.exports = {
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/storage.ts]: This should become an `export const setup =`

File: lib/renderer/extensions/web-navigation.ts [hunk 9]
```diff
@@ -1,18 +1,16 @@
...
 class WebNavigation {
-  constructor () {
-    this.onBeforeNavigate = new Event()
-    this.onCompleted = new Event()
+  private onBeforeNavigate = new Event()
+  private onCompleted = new Event()
 
-    ipcRendererInternal.on('CHROME_WEBNAVIGATION_ONBEFORENAVIGATE', (event, details) => {
+  constructor () {
+    ipcRendererInternal.on('CHROME_WEBNAVIGATION_ONBEFORENAVIGATE', (event: Event, details: any) => {
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/web-navigation.ts]: This is gonna use the wrong `Event` so I'm surprised TS allowed it.  You shouldn't have to define the types here, TS should infer them from `electron.d.ts`
```suggestion
    ipcRendererInternal.on('CHROME_WEBNAVIGATION_ONBEFORENAVIGATE', (event, details) => {
```

Comment (MarshallOfSound) [reply lib/renderer/extensions/web-navigation.ts]: Same below

Comment (codebytere) [reply lib/renderer/extensions/web-navigation.ts]: I assumed this too but it blows up with 

```
details has implicit type 'any'
```
etc

File: lib/renderer/extensions/web-navigation.ts [hunk 10]
```diff
@@ -1,23 +1,20 @@
...
+    ipcRendererInternal.on('CHROME_WEBNAVIGATION_ONCOMPLETED', (event: IpcMainEvent, details: any) => {
       this.onCompleted.emit(details)
     })
   }
 }
 
-exports.setup = () => {
-  return new WebNavigation()
-}
+exports.setup = () => new WebNavigation()
```
Comment (MarshallOfSound) [inline lib/renderer/extensions/web-navigation.ts]:
```suggestion
export const setup = () => new WebNavigation()
```

I would also question this method, seems a bit useless, why not just export the `class` 😆 (I know it was like this before but hey)

**Pull request merged**